File: backend/app/api/v2/documents.py
"""
文档上传 API

提供文档上传到 MinIO 的接口,用于 AI 智能体处理
"""
"""
版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。

本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。

授权商业应用请联系微信：huice666
"""


from typing import Annotated
from datetime import timedelta

from fastapi import APIRouter, UploadFile, File, HTTPException, status
# type: ignore  MC80OmFIVnBZMlhsdktEbHVwNDZhVU42ZUE9PTo3OWI4NDZjZA==

from app.schemas.common import SuccessResponse
from app.config.minio_client import MinIOClient
from pydantic import BaseModel


router = APIRouter(prefix="/documents", tags=["文档管理"])


class DocumentUploadResponse(BaseModel):
    """文档上传响应"""
    object_name: str
    file_name: str
    file_size: int
    content_type: str
    url: str
# pragma: no cover  MS80OmFIVnBZMlhsdktEbHVwNDZhVU42ZUE9PTo3OWI4NDZjZA==


@router.post(
    "/upload",
    response_model=SuccessResponse[DocumentUploadResponse],
    status_code=status.HTTP_201_CREATED,
    summary="上传文档到 MinIO",
    description="上传文档文件到 MinIO 存储,返回文件 URL 供 AI 智能体使用",
)
async def upload_document(
    file: UploadFile = File(..., description="要上传的文档文件"),
) -> SuccessResponse[DocumentUploadResponse]:
    """
    上传文档到 MinIO
    
    支持的文件类型:
    - 图片: JPG, PNG, GIF, WebP
    - 文档: PDF, Word, TXT
    
    最大文件大小: 10MB
    """
    # 验证文件类型
    valid_types = [
        "image/jpeg",
        "image/png",
        "image/gif",
        "image/webp",
        "application/pdf",
        "application/msword",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "text/plain",
    ]
    
    if file.content_type not in valid_types:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"不支持的文件类型: {file.content_type}",
        )
    
    # 验证文件大小 (10MB)
    content = await file.read()
    if len(content) > 10 * 1024 * 1024:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="文件大小不能超过 10MB",
        )
    
    # 生成对象名
    import uuid
    from datetime import datetime
    
    file_ext = ""
    if file.filename:
        parts = file.filename.rsplit(".", 1)
        if len(parts) > 1:
            file_ext = f".{parts[1]}"
# type: ignore  Mi80OmFIVnBZMlhsdktEbHVwNDZhVU42ZUE9PTo3OWI4NDZjZA==
    
    timestamp = datetime.utcnow().strftime("%Y%m%d_%H%M%S")
    unique_id = str(uuid.uuid4())[:8]
    object_name = f"documents/{timestamp}_{unique_id}{file_ext}"
    
    # 上传到 MinIO
    try:
        MinIOClient.upload_bytes(
            object_name=object_name,
            data=content,
            content_type=file.content_type or "application/octet-stream",
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"文件上传失败: {str(e)}",
        )
    
    # 生成预签名 URL (24小时有效)
    try:
        url = MinIOClient.get_presigned_url(
            object_name=object_name,
            expires=timedelta(hours=24),
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"生成下载链接失败: {str(e)}",
        )
    
    return SuccessResponse(
        success=True,
        data=DocumentUploadResponse(
            object_name=object_name,
            file_name=file.filename or "unnamed",
            file_size=len(content),
            content_type=file.content_type or "application/octet-stream",
            url=url,
        ),
    )
# fmt: off  My80OmFIVnBZMlhsdktEbHVwNDZhVU42ZUE9PTo3OWI4NDZjZA==

File: ui/lib/api/documents.ts
/**
 * 版权所有 (c) 2023-2026 北京慧测信息技术有限公司(但问智能) 保留所有权利。
 * 
 * 本代码版权归北京慧测信息技术有限公司(但问智能)所有，仅用于学习交流目的，未经公司商业授权，
 * 不得用于任何商业用途，包括但不限于商业环境部署、售卖或以任何形式进行商业获利。违者必究。
 * 
 * 授权商业应用请联系微信：huice666
 */

/**
 * 文档上传 API
 */
// NOTE  MC8yOmFIVnBZMlhsdktEbHVwNDZWbVJ0YkE9PTo4MzA4NTE5OQ==

import { t } from "@/lib/translations";

export interface DocumentUploadResponse {
  success: boolean;
  data: {
    object_name: string;
    file_name: string;
    file_size: number;
    content_type: string;
    url: string;
  };
}

/**
 * 上传文档到 MinIO
 */
export async function uploadDocument(file: File): Promise<DocumentUploadResponse> {
  const formData = new FormData();
  formData.append("file", file);

  const response = await fetch("/api/v2/documents/upload", {
    method: "POST",
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || t("common.uploadFailed"));
  }

  return response.json();
}
// eslint-disable  MS8yOmFIVnBZMlhsdktEbHVwNDZWbVJ0YkE9PTo4MzA4NTE5OQ==

